staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go

```diff
@@ -66,28 +70,61 @@ func newWithClock(authenticator authenticator.Token, cacheErrs bool, successTTL,
 	}
 }
 
+const (
+	// We have to cache for a non-zero amount of time in order to be able to share work.
+	minimumCacheTTL = 1 * time.Microsecond
+
+	// if it takes longer than this to perform the underlying auth call, it
+	// will still succeed, but other calls arriving after this time and
+	// before its completion will not share its result.
+	maxComputationTimeForJoining = 500 * time.Millisecond
+)
+
+func (a *cachedTokenAuthenticator) lookupFunc(ctx context.Context, token string) lrucache.ComputeFunc {
+	return func(abort <-chan time.Time) (value interface{}, ttl time.Duration) {
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		go func() {
+			select {
+			case <-abort:
+				cancel()
+			case <-ctx.Done():
+			}
+		}()
```

Comment:
It's probably simpler and less error prone to have goroutines waiting on `<-ce.wait` handle their own timeouts rather than requiring that lookupFuncs properly abort.